### PR TITLE
Update mace

### PR DESCRIPTION
**Pull request by Hanwen1018**
update al train to parallel tensor

### CODE REVIEW

File: mlptrain/training/active.py [hunk 1]
```diff
@@ -69,7 +70,8 @@ def train(mlp:               'mlptrain.potentials._base.MLPotential',
 
         init_configs: (gt.ConfigurationSet) A set of configurations from
                       which to start the active learning from
-
+        
+        torch: (bool) if want to share tensor between processes, set torch == True
```
Comment (tristan-j-wood) [inline mlptrain/training/active.py:74]: Slightly unclear what this means(?). Do you mean if you want to share tensor calculations between processes, use torch == True? Also need capitalisation at the start of the sentence and don't need to specify how to set it in the doc string.

**Review by t-young31 — COMMENTED** [review]
looks nice – thanks. In the interests of making this useful for potentially other MLP methods I would have the required context name as an MLP class attribute. i.e 

```python

class MLPotential(ABC):

    _mp_context_name = "fork"
```

and then in the `MACE` class override it as `"spawn"`. 

*** 

Just thinking that this is may be required in more than one place, and is really a property of the MLP method (opposed to a argument of a training function). And as an aside, is there a MACE wrapper (`MLPotential` subclass) to be added to this PR??

File: mlptrain/training/active.py (hunk 1 above)
Comment (Hanwen1018) [reply mlptrain/training/active.py]: I do not understand what you mean by tensor calculation.

**Comment by Hanwen1018** [comment]
Thank you both for your comments and suggestions

I am updating MACE to the package but it is not in this PR. I will doing
the other PR to wrap the MACE potential soon.

Yeah, I agree it is a property of the potential rather than a argument of
al train. Will add it to the potential property.

Thank you



Tom Young ***@***.***> 于2022年10月14日周五 15:16写道：

File: mlptrain/training/active.py (hunk 1 above)
Comment (tristan-j-wood) [reply mlptrain/training/active.py]: The doc string isn't clear to me what the bool represents or does, so needs to be clarified. What do you mean by share tensor between processes?

File: mlptrain/potentials/mace/mace.py [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+
+class MACE(MLPotential):
+
+    def _train_file(self,
+                    valid_fraction = 0.1,
+                    config_type_weights = '{"Default":1.0}' ,
+                    model = 'MACE',
+                    loss = 'weighted',
+                    hidden_irreps = '128x0e + 128x1o',
+                    batch_size = 10,
```
Comment (tristan-j-wood) [inline mlptrain/potentials/mace/mace.py:39]: Some of variables have spaces between the = and some don't. All need to have a space either side.
The variable type need Typing hints to be consistent with the rest of the code.
Some of the arguments are commented out(?)

File: mlptrain/potentials/mace/mace.py [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+                    #start_swa= 1200,
+                    ema = False,
+                    ema_decay=0.99,
+                    amsgrad = True,
+                    #max_num_epochs=800,
+                    restart_latest = True,
+                    save_cpu = True
+                    ) -> None:
+        """
+        Train this potential on the current training data is adjusted from run_train.py in mace package
```
Comment (tristan-j-wood) [inline mlptrain/potentials/mace/mace.py:54]: "Train this potential on the current training data is adjusted" - doesn't make sense to me

File: mlptrain/potentials/mace/mace.py [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+        https://github.com/ACEsuit/mace/tree/main/scripts
+        """
+
+        if len(self.training_data) <100:
```
Comment (tristan-j-wood) [inline mlptrain/potentials/mace/mace.py:58]: Please follow the PEP-8 guide for spacing

File: mlptrain/potentials/mace/mace.py [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+            max_num_epochs = 1000
+        elif 100<=len(self.training_data) <300:
```
Comment (tristan-j-wood) [inline mlptrain/potentials/mace/mace.py:60]: A comment on why these magic numbers are chosen is perhaps needed

File: mlptrain/potentials/mace/mace.py [hunk 6]
```diff
@@ -0,0 +1,556 @@
...
+
+
+@dataclasses.dataclass
+class SubsetCollection:
+    train: data.Configurations
+    valid: data.Configurations
+
+class MACE(MLPotential):
+
+    def _train_file(self,
```
Comment (tristan-j-wood) [inline mlptrain/potentials/mace/mace.py:33]: This is a huge function, probably worth splitting it up into smaller functions as it is difficult to follow what is going on

**Pull request closed without merging**